File: network/tools/ruby/echo_client.rb
## 2022-10-01 K.OHWADA
# echo_client.rb
# Messages to be sent and received 
# are separated by newline code

# https://ruby-doc.org/stdlib-2.1.0/libdoc/socket/rdoc/Socket.html

require 'socket'

PORT = 2000

sock = TCPSocket.new 'localhost', PORT

SEND_FLAGS = 0
RECV_LENGTH = 100
SLEEP = 1 # 1 sec

for num in 1..10 do
     msg = sprintf("%03d", num)
    sock.send(msg, SEND_FLAGS)
    if line = sock.recv(RECV_LENGTH)
        printf("r: %s \n", line)
    end
    sleep(SLEEP)
end
printf("close socket \n")
sock.close

File: network/tools/perl/echo_client.pl
## 2022-10-01 K.OHWADA

#  echo_client.pl
# Messages to be sent and received 
# are separated by newline code
# https://perlzemi.com/blog/20090406123890.html

# remove the trailing newline code
sub rtrim {
	my $val = shift;
	$val =~ s/\s*$//;
	return $val;
}

# use strict;
# use warnings;
use Socket;

# Host name to connect to
my $host = 'localhost';
# port number to connect to
my $port = 5000;

# Client
# 1. Create Socket
my $sock;
socket($sock, PF_INET, SOCK_STREAM, getprotobyname('tcp' ))
  or die "Cannot create socket: $!";

# 2. Create socket information
my $packed_remote_host = inet_aton($host)
  or die "Cannot pack $host: $!";

# Pack hostname and port number
my $sock_addr = sockaddr_in($port, $packed_remote_host)
  or die "Cannot pack $host:$port: $!";

# 3. Connect using socket
connect($sock, $sock_addr)
  or die "Cannot connect $host:$port: $!";

# 4. Write data
# Don't do write buffering.
my $old_handle = select $sock;
$| = 1; 
select $old_handle;

# the server does not receive without a linefeed code
my $format = "%03d\n";

my $data;

my $count;

# for loop ---
for($count = 0; $count < 10; $count++) {
    my $msg = sprintf( $format, $count );

    print $sock $msg;

# 5. read data
    if(my $rcv_data = <$sock> ) {
        my $data = rtrim($rcv_data);
        print "r: $data \n";
    }

# wait 1 sec
    sleep 1;
}
# for ---

# 6. Close socket
close $sock;

1;
